### ISSUE TITLE
KL-reweighting not done correctly in BNN code.

### ISSUE THREAD

**Issue by malharjajoo** [tensorflow/probability#719]
Hi,

In section 3.4 of this paper: [Weight Uncertainty in Neural Networks](https://arxiv.org/pdf/1505.05424.pdf), the KL divergence is re-weighted using "number of batches", but in the BNN example code [here](https://github.com/tensorflow/probability/blob/563cdd7bb46e3ce0bd069c713c581e7409f26de1/tensorflow_probability/examples/bayesian_neural_network.py#L264), the KL divergence is re-weighted using size of batch (when using stochastic gradient descent).

Does this seem like a bug ?

Thanks

**Comment by nbro** [comment]
@malharjajoo This question has been asked before on this issue tracker. See, for example, the issue I had opened: https://github.com/tensorflow/probability/issues/651. In the paper you mention, the authors actually mention more than one way of weighting the KL term of the ELBO.

**Comment by malharjajoo** [comment]
@nbro, thanks for that. I can see that the code is actually unclear, even for others.

Also, I think the KL is being divided by size of batch, because that [code](https://github.com/tensorflow/probability/blob/563cdd7bb46e3ce0bd069c713c581e7409f26de1/tensorflow_probability/examples/bayesian_neural_network.py#L161) is building a **batched** data pipeline using tf.data API. and when I ran through the [the line of code I mentioned earlier](https://github.com/tensorflow/probability/blob/563cdd7bb46e3ce0bd069c713c581e7409f26de1/tensorflow_probability/examples/bayesian_neural_network.py#L264) with a debugger, `mnist_data.train.num_examples`turns out to be =128 (which is default batch size in their code).

**Comment by nbro** [comment]
>  because that code is building a batched data pipeline using tf.data API. and when I ran through the the line of code I mentioned earlier with a debugger, `mnist_data.train.num_examples` turns out to be =128 (which is default batch size in their code).

I haven't checked it, but are you sure? I would double-check it. `mnist_data.train.num_examples` is a property of `train`, which is a member of the object `mnist_data`, which is returned by `mnist.read_data_sets`

https://github.com/tensorflow/probability/blob/563cdd7bb46e3ce0bd069c713c581e7409f26de1/tensorflow_probability/examples/bayesian_neural_network.py#L224

`mnist_data.train.num_examples` shouldn't be a placeholder or something that is filled during training. What makes you think that the input pipeline is modifying `mnist_data.train.num_examples`?
